### PR TITLE
fix(beam): GlobalCounters kernel count increasing when clearing l2

### LINKED ISSUE #4595: BEAM kernel count number is wrong
```
beam2  : 16 31 31 16 2 3 2 4 3 2        :   817.92 us  <  hc     : 4 31 31 32 4 3 3 4 2 2         :  1000.83 us
*** GPU        9 r_16_31_31_16_2_3_2_4_3_2              arg   3 mem  0.87 GB tm   1244.89us/     4.99ms (  113.83 GFLOPS,   24.03 GB/s)
   0.00s:                 from   1 ->   1 actions  512   32   32   32   12
   4.62s:      7940.00 us from  30 ->  30 actions  512    2   32   32   16   12
  13.97s:      3760.83 us from  59 ->  59 actions  512    2   32    8   16   12    4
  23.99s:      2318.33 us from  56 ->  56 actions  512    2   32   16   12    4    8
  35.39s:      1741.56 us from  47 ->  47 actions  512   32    4   16    2   12    4    2
  43.81s:      1741.56 us from  41 ->  41 actions  512   32    4   16    2   12    4    2
beam2  : 512 32 4 16 2 12 4 2           :  1778.44 us  <  hc     : 512 32 2 16 8 4 12             :  3710.63 us
*** GPU       16 r_512_32_4_16_2_12_4_2                 arg   3 mem  0.96 GB tm   1835.73us/     6.83ms (  274.18 GFLOPS,   49.42 GB/s)
```

Kernel number jumps from 9 to 16, it should only go up by one (like from 9 to 10)

### PR DESCRIPTION
**Pull request by Leikoe**
fixes #4595 by not updating stats when clearing l2 cache in beam search. Added a NOSTATS env var to prevent stats update with a context when some operations need to be ran without updating framework's stats.

### CODE REVIEW

**Comment by Leikoe** [comment]
Using the already implemented do_update_stats is better and seems more reusable

**Comment by geohot** [comment]
Cool! Can you add a test to prevent regression?

**Comment by Leikoe** [comment]
> Cool! Can you add a test to prevent regression?

yes!

File: test/test_search.py [hunk 1]
```diff
@@ -66,5 +66,22 @@ def test_get_linearizer_actions(self):
...
+    dev = Device[Device.DEFAULT]
+    assert dev.compiler is not None
+
+    si = [i for i in create_schedule([Tensor([1,2,3,4]).add(1).lazydata]) if i.ast[0].op not in LoadOps][0]
+    rawbufs = _ensure_buffer_alloc(bufs_from_lin(lin:=Linearizer(*si.ast)))
+    var_vals = {k:(k.max+k.min)//2 for k in lin.ast[0].vars()}
+    p = lin.to_program()
+
+    kernel_count = GlobalCounters.kernel_count
+    _time_program(p, dev.compiler.compile(p.src), var_vals, rawbufs,
```
Comment (chenyuxyz) [inline test/test_search.py:82]: prefer not to test internal functions directly (function name that start with `_`). can you use `time_linearizer` instead? also kernel count not changing may due to nothing is being run, i'd test the count increased by exactly one after running one kernel

File: test/test_search.py [hunk 2]
```diff
@@ -66,5 +66,22 @@ def test_get_linearizer_actions(self):
     if Opt(OptOps.GROUPTOP, 0, 0) in actions:
       assert len([x for x in lins if x.applied_opts[0] == Opt(OptOps.GROUPTOP, axis=0, amt=3)]) == 0, "did not de-dup GROUPTOP"
 
+  def test_kernel_count(self):
```
Comment (chenyuxyz) [inline test/test_search.py:69]: fine to skip "NV" per #4607

Comment (nimlgen) [reply test/test_search.py:69]: fyi fixed in master, should pass now

File: test/test_search.py (hunk 1 above)
Comment (Leikoe) [reply test/test_search.py]: I switched to testing `time_linearizer`. Though I don't understand the "running one kernel part". Is time_linearizer supposed to run kernels which count towards kcount ?

Comment (chenyuxyz) [reply test/test_search.py]: it was 2 separate asks (1) don't call internal function in test, (2) test one kernel ran and the kernel count increased by 1 (not 7 or 10 in search).

**Comment by geohot** [comment]
Thanks!

File: test/test_search.py [hunk 3]
```diff
@@ -66,6 +66,19 @@ def test_get_linearizer_actions(self):
...
+    Ensure that the kernel count is not incremented by time_linearizer when clearing l2
+    """
+    # ast of Tensor.zeros(16).contiguous().realize()
+    ast = LazyOp(op=BufferOps.STORE, src=(LazyOp(op=BufferOps.CONST, src=(), arg=ConstBuffer(val=0.0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(16,), strides=(0,), offset=0, mask=None, contiguous=False),)))),), arg=MemBuffer(idx=0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(16,), strides=(1,), offset=0, mask=None, contiguous=True),))))  # noqa: E501
+    lin = Linearizer(ast)
+    bufs = bufs_from_lin(lin)
+
+    kernel_count = GlobalCounters.kernel_count
+    time_linearizer(lin, bufs, allow_test_size=False, cnt=2, disable_cache=True, clear_l2=True)
+    assert GlobalCounters.kernel_count == kernel_count, "kernel count was incremented by time_linearizer"
```
Comment (chenyuxyz) [inline test/test_search.py:80]: hmm did it increment?

Comment (Leikoe) [reply test/test_search.py:80]: its not supposed to

**Pull request merged**